### ISSUE TITLE
Arango crashed hard executing very simple query on windows

### ISSUE THREAD

**Issue by ctmakro** [arangodb/arangodb#1823]
Below are the AQL used.
````
//for f in forums
//let f = document(forums,'default')

    let count_posts = (
      for t in threads
      filter t.fid == f._key
      collect aggregate total = sum(t.count)
      return total
    )[0]

    let count_threads = (
      for t in threads
      filter t.fid == f._key
      collect with count into total
      return total
    )[0]

    let count_posts_today = (
      for t in threads
      filter t.fid == f._key && t.tlm > date_now()-86400*1000
      collect aggregate total = sum(t.count_today)
      return total
    )[0]

    update f with {count_posts,count_posts_today,count_threads} in forums
````
by using `let f = document(forums,'default')` everything's okay;
by using `for f in forums` (forums is a collection with only 3 documents) the ArangoDB service wil immediately crash(100% chance), leaving me an ECONNRESET.

NodeJS 4.x / Win7 64-bit
Same problem in both 2.8.3 and 2.8.7
Same problem using both arangojs driver, and AQL editor

This is very annoying since I'm meeting deadlines for [some company project](https://github.com/lovetheory/nkc2) .

**Comment by fceller** [comment]
Hi ctmakro, I will try to reproduce.

**Comment by fceller** [comment]
@ctmakro can you send me test data (hackers (at) arangodb.org). For example, a dump of your dataset? I tried to create a collection forums with 3 entries. This does not reproduce the crash. Therefore, it would be good to get a test set.

**Comment by ctmakro** [comment]
I did more comparison tests.

Problematic:
````
let count_posts_today = (
      for t in threads
      filter t.fid == f._key && t.tlm > date_now()-86400*1000
      collect aggregate total = sum(t.count_today)
      return total
    )[0]
````

Perfect:
````
let count_posts_today = (
      for t in threads
      filter t.fid == f._key
      collect aggregate total = sum(t.count_today)
      return total
    )[0]
````

I will send you a dump within half an hour.

**Comment by ctmakro** [comment]
just sent the email.
expect lots of Chinese characters :)
BTW you people at Arango are making great software. Really looking foward for a production ready 3.0.

**Comment by ctmakro** [comment]
I did even more comparison test.

Below will work
````
for f in forums
    
    let count_posts_today = (
      for t in threads
      filter t.fid == f._key && t.tlm > DATE_NOW() 
      collect aggregate total = sum(t.count_today)
      return total
    )[0]
    
    update f with {count_posts_today} in forums
    return NEW
````

Below will crash - by adding a minus operator and an integer.
` filter t.fid == f._key && t.tlm > DATE_NOW() - 86400000`

**Comment by ctmakro** [comment]
After digging quite a while, guess what:

Below will crash, but IT WILL SURVIVE IF 10000000 is changed to 1000.
````
    for f in forums
    
    let count_posts_today = (
      for t in threads
      filter t.fid == f._key && t.tlm > DATE_NOW() - 10000000 
      collect aggregate total = sum(t.count_today)
      return total
    )[0]
    
    update f with {count_posts_today} in forums
    return NEW
````
some kind of overflow? hope that won't affect server security.

**Comment by fceller** [comment]
No, it looks more like a bug in the query engine, need to investigate.

**Comment by ctmakro** [comment]
I tried using bind parameters and the problem still exists.
By testing using the divide-by-2 method, I found that the interger will cause a crash when
greater than 9597151
or
less than 9395727.

That's the weirdest bug since my NoSQL journey.

**Comment by fceller** [comment]
@ctmakro there is a bug in the query engine, which causes this crash. However, if I understand the intent correctly, the query should be rewritten as follows:

```
for f in forums
    let posts = (
      for t in threads
      filter t.fid == f._key
      return t.count
    )
    
    let count_posts = sum(posts)
    let count_threads = length(posts)

    let count_posts_today = sum(
      for t in threads
      filter t.fid == f._key && t.tlm > date_now()-86400*1000
      return t.count_today
    )

    update f with {count_posts,count_posts_today,count_threads} in forums
```

The "collect into" should not be necessary, because you want to count given the number of matching documents or sum up the matching threads.

**Comment by ctmakro** [comment]
More tests. The critical value shifted with time. Later i found that Arango crashes when at least one 'thread' has `t.tlm > some_timestamp`.
Below will crash:
````
    for f in forums
    
    let count_posts_today = (
      for t in threads
      filter t.fid == f._key && t.tlm > 1463000000000
      collect aggregate total = sum(t.count_today)
      return total
    )[0]
    
    update f with {count_posts_today} in forums
    return NEW
````

**Comment by fceller** [comment]
You would need the "count into", if you you want to avoid the outer loop ("for f in forums"):

```
for t in threads
  collect key = t.fid into c = t.count
  return [key, sum(c), length(c)]
```

**Comment by fceller** [comment]
@ctmakro the query should be

```
for f in forums

    let count_posts_today = sum(
      for t in threads
      filter t.fid == f._key && t.tlm > 1463000000000
      return t.count_today
    )

    update f with {count_posts_today} in forums
    return NEW
```

Can you try this? Please also check my remark above.

**Comment by ctmakro** [comment]
You are correct: that solved my problem.
Below worked flawlessly:
````
    for f in forums
    
    let count_posts_today = sum(
      for t in threads
      filter t.fid == f._key && t.tlm > 1463000000000
      return t.count_today
    )
    
    update f with {count_posts_today} in forums
    return NEW
````
Really appreciated that. I should read the documentation more carefully.

**Comment by fceller** [comment]
Good to hear. I will leave the issue open, because ArangoDB should not crash.

**Comment by jsteemann** [comment]
The crash has been fixed in the 2.8 branch. The fix will be included in version 2.8.10.
